src/pybind/mgr/mds_autoscaler/module.py

```diff
@@ -0,0 +1,139 @@
+        # standbys are not grouped by filesystems in fs_map
+        # available = standby_replay + standby_active
+        assert fs_map is not None
+        total = 0
+        daemon_names = {
+            d.name() for d in daemons
+        }
+        for sb in fs_map['standbys']:
+            full_name = "mds.{}".format(sb['name'])
+            if full_name in daemon_names:
+                total += 1
+        return total
+
+    def get_current_active_count(self, fs_map: dict, fs_name: str) -> int:
+        assert fs_map is not None
+        for fsys in fs_map['filesystems']:
+            if fsys.get('mdsmap').get('fs_name') == fs_name:
+                return len(fsys.get('mdsmap').get('up'))
+        assert False
+
+    def get_required_max_mds(self, fs_map: dict, fs_name: str) -> int:
+        assert fs_map is not None
+        for fsys in fs_map['filesystems']:
+            if fsys.get('mdsmap').get('fs_name') == fs_name:
+                return fsys.get('mdsmap', {}).get('max_mds')
+        assert False
+
+    def verify_and_manage_mds_instance(self, fs_map: dict, fs_name: str):
+        assert fs_map is not None
+
+        try:
+            daemons = self.get_daemons(fs_name)
+            standbys_required = self.get_required_standby_count(fs_map, fs_name)
+            standbys_current = self.get_current_standby_count(fs_map, fs_name, daemons)
+            active = self.get_current_active_count(fs_map, fs_name)
+            max_mds_required = self.get_required_max_mds(fs_map, fs_name)
+
+            self.log.debug(f"fs_name:{fs_name} "
+                           f"standbys_required:{standbys_required}, "
+                           f"standbys_current:{standbys_current}, "
+                           f"active:{active}, "
+                           f"max_mds_required:{max_mds_required}")
+
+            total_current = standbys_current + active
+            total_required = max_mds_required + standbys_required
+            self.log.info(f"fs:{fs_name} total_required:{total_required}, total_current:{total_current}")
+
+            if total_required < total_current:
+                self.log.info(f"fs:{fs_name}, killing {total_current - total_required} standby mds ...")
+            elif total_required > total_current:
+                self.log.info(f"fs:{fs_name}, spawning {total_required - total_current} standby mds ...")
+            else:
+                self.log.debug(f"fs:{fs_name} no change to mds count")
+                return
+
+            newspec = self.update_daemon_count(fs_name, total_required)
+
+            self.log.info(f"fs:{fs_name}, new placement count:{newspec.placement.count}")
+
+            completion = self.apply_mds(newspec)
+            self._orchestrator_wait([completion])
+            orchestrator.raise_if_exception(completion)
+        except (orchestrator.OrchestratorError, AssertionError) as e:
+            self.log.debug(f"fs:{fs_name} exception while verifying mds status: {e!r}")
```

Comment:
```suggestion
        except orchestrator.OrchestratorError as e:
            self.log.exception(f"fs:{fs_name} exception while verifying mds status: {e}")
```

Reply:
Don't catch `AssertionError`, we want the module to blow up if that hits.

Reply:
fixed